### PR TITLE
Replace signed in user for identity.requester

### PR DESCRIPTION
**Pull request by linoman**
Continuing pending work from https://github.com/grafana/grafana/pull/73750

The work was merged and had to be reverted. This new PR addresses previous comments from the PR that were left unattended.

**What is this feature?**

This PR will replace the user of `*user.SignedInUser` for the `identity.Requester` interface.

On top of those changes, I also "renamed" some status codes from numbers to constants for the HTTP module.

There are very little changes to the actual code other than validation of the namespace when needed. If the validation is failed, then logs will be used.

**Why do we need this feature?**

[Add a description of the problem the feature is trying to solve.]

**Who is this feature for?**

[Add information on what kind of user the feature is for.]

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/identity-access-team/issues/272

**Special notes for your reviewer:**

The build might fail if the enterprise PR is not merged first.

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: public/app/core/components/AppChrome/AppChromeService.tsx [hunk 1]
```diff
@@ -35,14 +35,14 @@ export class AppChromeService {
     layout: PageLayoutType.Canvas,
   });
 
-  public setMatchedRoute(route: RouteDescriptor) {
+  setMatchedRoute(route: RouteDescriptor) {
```
Comment (ryantxu) [inline public/app/core/components/AppChrome/AppChromeService.tsx:38]: Are all the typescript changes unrelated?

Comment (linoman) [reply public/app/core/components/AppChrome/AppChromeService.tsx:38]: They are not. I must've run a linter somewhere. I'll revert this.

Comment (linoman) [reply public/app/core/components/AppChrome/AppChromeService.tsx]: Fixed.

File: pkg/services/folder/folderimpl/folder.go [hunk 2]
```diff
@@ -280,7 +281,12 @@ func (s *Service) Create(ctx context.Context, cmd *folder.CreateFolderCommand) (
 	dashFolder.SetUID(trimmedUID)
 
 	user := cmd.SignedInUser
-	userID := user.UserID
+	namespaceID, userIDstr := user.GetNamespacedID()
```
Comment (Jguer) [inline pkg/services/folder/folderimpl/folder.go:284]: check namespace, if API key set user ID to 0 if you want to keep legacy support for API keys.

File: pkg/services/folder/folderimpl/folder.go [hunk 3]
```diff
@@ -763,12 +769,18 @@ func (s *Service) BuildSaveDashboardCommand(ctx context.Context, dto *dashboards
 		}
 	}
 
+	namespaceID, userIDstr := dto.User.GetNamespacedID()
+	userID, err := identity.IntIdentifier(namespaceID, userIDstr)
```
Comment (Jguer) [inline pkg/services/folder/folderimpl/folder.go:773]: check namespace, if API key set user ID to 0 if you want to keep legacy support for API keys.

File: pkg/services/guardian/guardian.go [hunk 4]
```diff
@@ -216,32 +216,36 @@ func (g *dashboardGuardianImpl) logHasPermissionResult(permission dashboards.Per
...
+
 	return hasPermission, err
 }
 
 func (g *dashboardGuardianImpl) checkACL(permission dashboards.PermissionType, acl []*dashboards.DashboardACLInfoDTO) (bool, error) {
-	orgRole := g.user.OrgRole
 	teamACLItems := []*dashboards.DashboardACLInfoDTO{}
 
 	for _, p := range acl {
+		userID, err := identity.IntIdentifier(g.user.GetNamespacedID())
```
Comment (Jguer) [inline pkg/services/guardian/guardian.go:236]: not guaranteed to be a user ID

Comment (Jguer) [reply pkg/services/guardian/guardian.go:236]: check namespace, if API key set user ID to 0 if you want to keep legacy support for API keys.

Comment (Jguer) [reply pkg/services/guardian/guardian.go:236]: Also allow render service with user ID 0 otherwise dashboards won't be rendered to PDF

Comment (linoman) [reply pkg/services/guardian/guardian.go]: This function has been removed after merge.

File: pkg/api/dashboard.go [hunk 5]
```diff
@@ -396,19 +421,26 @@ func (hs *HTTPServer) PostDashboard(c *contextmodel.ReqContext) response.Respons
...
-	cmd.UserID = c.UserID
+
+	namespaceID, userIDstr := c.SignedInUser.GetNamespacedID()
+	userID, err := identity.IntIdentifier(namespaceID, userIDstr)
+	if err != nil {
+		hs.log.Warn("Error while parsing user ID", "namespaceID", namespaceID, "userID", userIDstr)
+	}
+
+	cmd.OrgID = c.SignedInUser.GetOrgID()
+	cmd.UserID = userID
```
Comment (Jguer) [inline pkg/api/dashboard.go:432]: namespace validation

File: pkg/services/user/identity.go [hunk 6]
```diff
@@ -51,6 +54,16 @@ func (u *SignedInUser) ToUserDisplayDTO() *UserDisplayDTO {
 	}
 }
 
+// Static function to parse a requester into a UserDisplayDTO
+func NewUserDisplayDTOFromRequester(requester identity.Requester) (*UserDisplayDTO, error) {
+	userID, _ := identity.IntIdentifier(requester.GetNamespacedID())
```
Comment (Jguer) [inline pkg/services/user/identity.go:59]: if it's not a service account or user, set to 0

File: pkg/services/dashboards/service/dashboard_service.go [hunk 7]
```diff
@@ -199,6 +208,22 @@ func (dr *DashboardServiceImpl) BuildSaveDashboardCommand(ctx context.Context, d
...
 
+func validateNamespace(user identity.Requester) (string, string, error) {
+	namespaceID, userIDstr := user.GetNamespacedID()
+	if namespaceID != identity.NamespaceUser && namespaceID != identity.NamespaceServiceAccount {
+		return "", "", errutil.BadRequest("account doesn't belong to the user or service namespace")
+	}
+	return namespaceID, userIDstr, nil
+}
+
+func resolveUserID(namespaceID, idenfitier string, log log.Logger) int64 {
```
Comment (Jguer) [inline pkg/services/dashboards/service/dashboard_service.go:219]: suggestion call validateNamespace from resolveUserID instead

File: pkg/api/dashboard_test.go [hunk 8]
```diff
@@ -174,6 +175,7 @@ func TestHTTPServer_GetDashboard_AccessControl(t *testing.T) {
 		require.NoError(t, err)
 
 		assert.Equal(t, http.StatusOK, res.StatusCode)
+		fmt.Println(res.Body)
```
[the remark this answers is not in the document]
Comment (linoman) [reply pkg/api/dashboard_test.go:178]: 🤦‍♂️

File: pkg/services/dashboards/service/dashboard_service.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (linoman) [reply pkg/services/dashboards/service/dashboard_service.go]: I merge the 2 functions in 1

File: pkg/services/user/identity.go [hunk 9]
```diff
@@ -51,6 +54,20 @@ func (u *SignedInUser) ToUserDisplayDTO() *UserDisplayDTO {
 	}
 }
 
+// Static function to parse a requester into a UserDisplayDTO
+func NewUserDisplayDTOFromRequester(requester identity.Requester) (*UserDisplayDTO, error) {
+	namespaceID, identifier := requester.GetNamespacedID()
+	if namespaceID != identity.NamespaceUser && namespaceID != identity.NamespaceServiceAccount {
+		identifier = "0"
+	}
+	userID, _ := identity.IntIdentifier(namespaceID, identifier)
```
Comment (Jguer) [inline pkg/services/user/identity.go:63]: return err in case there is a failure in auth broker

**Review by Jguer — APPROVED** [review]

**Comment by torkelo** [comment]
@Jguer @linoman this look to break anonymous access , incident on play.grafana.org now but can replicate locally by just enabling anonymous access

File: pkg/api/dashboard.go [hunk 10]
```diff
@@ -515,12 +556,22 @@ func (hs *HTTPServer) postDashboard(c *contextmodel.ReqContext, cmd dashboards.S
 // 401: unauthorisedError
 // 500: internalServerError
 func (hs *HTTPServer) GetHomeDashboard(c *contextmodel.ReqContext) response.Response {
-	prefsQuery := pref.GetPreferenceWithDefaultsQuery{OrgID: c.OrgID, UserID: c.SignedInUser.UserID, Teams: c.Teams}
+	namespaceID, userIDstr := c.SignedInUser.GetNamespacedID()
+
+	if namespaceID != identity.NamespaceUser {
+		return response.Error(http.StatusBadRequest, "User does not belong to a user namespace", nil)
+	}
+
```
Comment (torkelo) [inline pkg/api/dashboard.go:564]: this is a strange check, why is this check here in GetHomeDashboard? it looks like a very generic middleware / auth check? 

Oh and it breaks anonymous access

Comment (Jguer) [reply pkg/api/dashboard.go:564]: https://github.com/grafana/grafana/pull/74397 this should cover a lot of the issues with anonymous access but I'll go through this one restoring ID 0 being passed to stores until we can have better checks on the store

**Pull request merged**